soapui/src/main/java/com/eviware/soapui/impl/wsdl/testcase/WsdlProjectRunner.java

```diff
@@ -153,85 +187,125 @@ private WsdlTestSuiteRunner runTestSuite(WsdlTestSuite testSuite, boolean async)
+            if (listeners == null || listeners.length == 0) {
+                return;
+            }
 
-        for (int i = 0; i < listeners.length; i++) {
-            listeners[i].afterTestSuite(this, getRunContext(), testSuiteRunner);
+            for (int i = 0; i < listeners.length; i++) {
+                listeners[i].afterTestSuite(this, getRunContext(), testSuiteRunner);
+            }
+        } finally {
+            lock.unlock();
         }
     }
 
     private void notifyBeforeRunTestSuite(TestSuite testSuite) {
-        if (listeners == null || listeners.length == 0) {
-            return;
-        }
+        try {
+            lock.lock();
+            if (listeners == null || listeners.length == 0) {
+                return;
+            }
 
-        for (int i = 0; i < listeners.length; i++) {
-            listeners[i].beforeTestSuite(this, getRunContext(), testSuite);
+            for (int i = 0; i < listeners.length; i++) {
+                listeners[i].beforeTestSuite(this, getRunContext(), testSuite);
+            }
+        } finally {
+            lock.unlock();
         }
     }
 
     public List<TestSuiteRunner> getResults() {
-        return Arrays.asList(finishedRunners.toArray(new TestSuiteRunner[finishedRunners.size()]));
-    }
-
-    protected void finishRunner(TestSuiteRunner testRunner) {
-        notifyAfterRunTestSuite(testRunner);
-
-        activeRunners.remove(testRunner);
-        finishedRunners.add(testRunner);
-
-        testRunner.getTestSuite().removeTestSuiteRunListener(internalTestRunListener);
-
-        if (activeRunners.isEmpty()) {
-            updateStatus();
-
-            synchronized (activeRunners) {
-                activeRunners.notify();
-            }
+        try {
+            lock.lock();
+            return Arrays.asList(finishedRunners.toArray(new TestSuiteRunner[finishedRunners.size()]));
+        } finally {
+            lock.unlock();
         }
     }
 
     private class InternalTestSuiteRunListener extends TestSuiteRunListenerAdapter {
         @Override
         public void afterRun(TestSuiteRunner testRunner, TestSuiteRunContext runContext) {
-            finishRunner(testRunner);
+            try {
+                while (!running) {
+                    try {
+                        Thread.sleep(100L);
+                    } catch (InterruptedException cause) {
+                        cause.printStackTrace();
+                    }
+                }
+
+                try {
+                    lock.lock();
+                    notifyAfterRunTestSuite(testRunner);
+
+                    finishedRunners.add(testRunner);
+
+                    testRunner.getTestSuite().removeTestSuiteRunListener(internalTestRunListener);
+
+
+                } finally {
+                    lock.unlock();
+                }
+            } finally {
+                runCount.decrementAndGet();
+                if (runCount.get() == 0) {
+                    updateStatus();

```

Comment:
The query above looks to call `updateStatus()` once at the end. But this code does not guaranties that `updateStatus()` is called once. I suggest
`if (runCount.decrementAndGet() == 0) {
        updateStatus();
}`